packages/context/src/context.ts

```diff
@@ -357,9 +375,12 @@ export class Context extends EventEmitter {
     this.registry.set(key, binding);
     if (existingBinding !== binding) {
       if (existingBinding != null) {
-        this.emit('unbind', existingBinding, this);
+        this.emit('unbind', existingBinding, this, 'unbind');
+        this.removeTagIndexForBinding(existingBinding);
```

Comment:
I am quite unhappy that you are modifying so many places to deal with tag indexing. I feel you are mixing implementation of different concerns in a single method/function, which make the codebase difficult to understand and modify.

Please find a way how to implement tag indexing as a de-coupled piece using events and/or ContextObserver to be notified about the relevant changes.

I am ok to include tag index inside the Context class, I just want the low-level implementation bits to live outside of Context.